tests/test_app.py

```diff
@@ -1381,3 +1384,232 @@ async def test_bulk_specification_fail_multiple_specs_per_type_excel():
+    async with AppClient(config) as cli:
+        resp = await cli.post("write_bulk_specification/", data="foo")
+        js = await resp.json()
+        assert js == {"error": "Required content-type is application/json"}
+        assert resp.status == 415
+
+
+async def test_write_bulk_specification_fail_no_content_length():
+    async with AppClient(config) as cli:
+        resp = await cli.post(
+            "write_bulk_specification/", headers={"content-type": "application/json"})
+        js = await resp.json()
+        assert js == {"error": "The content-length header is required and must be > 0"}
+        assert resp.status == 411
+
+
+async def test_write_bulk_specification_fail_large_input():
+    async with AppClient(config) as cli:
+        resp = await cli.post("write_bulk_specification/", json="a" * (1024 * 1024 - 2))
+        txt = await resp.text()
+        # this seems to be a built in (somewhat inaccurate) server feature
+        assert txt == "Maximum request body size 1048576 exceeded, actual body size 1048576"
+        assert resp.status == 413
+
+
+async def test_write_bulk_specification_fail_not_dict():
+    async with AppClient(config) as cli:
+        resp = await cli.post("write_bulk_specification/", json=["foo"])
+        js = await resp.json()
+        assert js == {"error": "The top level JSON element must be a mapping"}
+        assert resp.status == 400
+
+
+async def test_write_bulk_specification_fail_no_output_dir():
+    async with AppClient(config) as cli:
+        resp = await cli.post("write_bulk_specification/", json={})
+        js = await resp.json()
+        assert js == {"error": "output_directory is required and must be a string"}
+        assert resp.status == 400
+
+
+async def test_write_bulk_specification_fail_wrong_type_for_output_dir():
+    async with AppClient(config) as cli:
+        resp = await cli.post("write_bulk_specification/", json={"output_directory": 4})
+        js = await resp.json()
+        assert js == {"error": "output_directory is required and must be a string"}
+        assert resp.status == 400
+
+
+async def test_write_bulk_specification_fail_no_file_type():
+    async with AppClient(config) as cli:
+        resp = await cli.post("write_bulk_specification/", json={"output_directory": "foo"})
+        js = await resp.json()
+        assert js == {"error": "Invalid output_file_type: None"}
+        assert resp.status == 400
+
+
+async def test_write_bulk_specification_fail_wrong_file_type():
+    async with AppClient(config) as cli:
+        resp = await cli.post("write_bulk_specification/", json=
+            {"output_directory": "foo", "output_file_type": "XSV"})
+        js = await resp.json()
+        assert js == {"error": "Invalid output_file_type: XSV"}
+        assert resp.status == 400
+
+
+async def test_write_bulk_specification_fail_invalid_type_value():
+    async with AppClient(config) as cli:
+        resp = await cli.post("write_bulk_specification/", json=
```

Comment:
might be able to refactor the first 3 lines of each of the tests to go into a helper function

Reply:
I did this for all but the first 3 of this style of test, which had unusual inputs or outputs